### ISSUE TITLE
How to set state of multiple providers at the same time?

### ISSUE THREAD

**Issue by vin-fandemand** [rrousselGit/river_pod#488]
I have multiple providers used by many submodules. Lets say they are Provider1,Provider2,Provider3. Sometimes I might want to set state of Provider2 when I am setting state of Provider1. Now as they are two different state changes and are executed one after the other, the screen becomes laggy or sometimes interferes with state building of the screen due to multiple updates. How do I prevent this ? How do I combine the state change of Provider1 and Provider2 and ensure it builds the screen as if the change happened only once ?

**Comment by rrousselGit** [comment]
Hi

I will need a bit more context. Could you describe the lag issue and your architecture more concretely?

**Comment by vin-fandemand** [comment]
[https://github.com/vin-fandemand/riverpodtest](url) has a sample app.
I was not able to recreate the issue with the simple example.
Here I have three services - `Service1`, `Service2`, and `Service3`. There are three main questions I have

1. How to initialize the widget - So when the widget loads, I want to initialize the services with some value (mostly the previous state from persistent storage for eg). For this, I created an init method which I call from `initState` or `didUpdateWidget` methods of the widget. Many times, when I do this, I get 

> widget cannot be marked as needing to build because the framework is already in the process of building widgets. A widget can be marked as needing to be built during the build phase only if one of its ancestors is currently building. 

So usually I just move the initialization of the service to a parent widget, instead of doing it inside the child's  `initState` or `didUpdateWidget` methods. How can I do it in the child itself without interfering with the state build ? Is there a better way to do this?

2. How to combine state changes for multiple providers - In the example, you can see that `Service1` has a method called `incrementCounter` which also makes some updates in `Service2` as well. As I understand, each of these state changes in the services would notify a consumer in the widget tree. Is there a way to combine such changes together so that all the consumers are notified in one go ?
3. When to properly do Autodispose or dispose ? Is there a best practice for that ?

**Comment by iamarnas** [comment]
@vin-fandemand Hi,
 I have never tried to combine providers in this way like in your example, but I feel you have laggy because you combine your providers over context. It's bad practice with `Riverpod`. You should combine your providers over `ProviderReference`

#### BAD:
```dart
class Service1 extends StateNotifier<ServiceState1> {
  Service1() : super(ServiceState1());

  void init(int field, BuildContext context) {
     state = state.copyWith(serviceField1: field);   
     context.read(service2.notifier).init(2);   
     context.read(service2.notifier).init(6);     
   }
  //.....
}
```
#### GOOD:
```dart
final service1 = StateNotifierProvider<Service1, ServiceState1>((ref) {
   return Service1(ref);   
 });

class Service1 extends <StateNotifier<ServiceState1>> {
  Service1(this.ref) : super(ServiceState1());

  final ProviderReference ref;

  void init(int field, BuildContext context) {
     state = state.copyWith(serviceField1: field);   
     ref.read(service2.notifier).init(2);   
     ref.read(service2.notifier).init(6);     
   }
  //.....
}
```
[More about how to combine providers.](https://riverpod.dev/docs/concepts/combining_providers)

I personally recommend avoid combine so many providers in one place. Because if you dispose one of the above provider in your providers tree then will be disposed all the rest to the bottom.

**Comment by vin-fandemand** [comment]
can you also clarify if I watch more than one provider in a consumer?

**Comment by iamarnas** [comment]
@vin-fandemand inside `Consumer` you can watch or read how many providers you want. `Consumer` are build for that.

**Comment by julienlebren** [comment]
You can also use `ProviderListener` in the top-level Widget of your page to listen the state of your `Service1` and init `Service2` and `Service3` when it changes.

```dart
class Service1 extends StateNotifier<ServiceState1> {
  Service1() : super(ServiceState1());
}

final service1 = StateNotifierProvider<Service1, ServiceState1>((ref) {
   return Service1(ref);   
});

final service2 = StateProvider<Service2?>((ref) => null);
final service3 = StateProvider<Service3?>((ref) => null);

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ProviderListener<ServiceState1>(
      provider: service1,
      onChange: (context, service1) {
         context.read(service2).state = Service2();
         context.read(service3).state = Service3();
        }
      },
      child: Text("Hello World"),
    );
  }
}
```

**Comment by vin-fandemand** [comment]
> You can also use `ProviderListener` in the top-level Widget of your page to listen the state of your `Service1` and init `Service2` and `Service3` when it changes.
> 
> ```dart
> class Service1 extends StateNotifier<ServiceState1> {
>   Service1() : super(ServiceState1());
> }
> 
> final service1 = StateNotifierProvider<Service1, ServiceState1>((_) => Service1());
> final service2 = StateProvider<Service2?>((ref) => null);
> final service3 = StateProvider<Service3?>((ref) => null);
> 
> class HomePage extends StatelessWidget {
>   const HomePage({Key? key}) : super(key: key);
> 
>   @override
>   Widget build(BuildContext context) {
>     return ProviderListener<ServiceState1>(
>       provider: service1,
>       onChange: (context, service1) {
>          context.read(service2).state = Service2();
>          context.read(service3).state = Service3();
>         }
>       },
>       child: Text("Hello World"),
>     );
>   }
> }
> ```

Yes thanks. It actually also depends on when you want to listen. Like for example, you would want to change the state of service2 after you called a particular method in service1.

**Comment by julienlebren** [comment]
It can be easily achieved by using `@freezed` class to generate your `ServiceState1`.
Then on `ProviderListener` you handle it like this:

```dart
state.maybeWhen(
  method1: () => doSomething1(),
  method2: () => doSomething2(),
  orElse: () => null,
);

**Comment by vin-fandemand** [comment]
is there a way to check if a service is in context without using `read()` ?

**Comment by vin-fandemand** [comment]
> [https://github.com/vin-fandemand/riverpodtest](url) has a sample app.
> I was not able to recreate the issue with the simple example.
> Here I have three services - `Service1`, `Service2`, and `Service3`. There are three main questions I have
> 
> 1. How to initialize the widget - So when the widget loads, I want to initialize the services with some value (mostly the previous state from persistent storage for eg). For this, I created an init method which I call from `initState` or `didUpdateWidget` methods of the widget. Many times, when I do this, I get
> 
> > widget cannot be marked as needing to build because the framework is already in the process of building widgets. A widget can be marked as needing to be built during the build phase only if one of its ancestors is currently building.
> 
> So usually I just move the initialization of the service to a parent widget, instead of doing it inside the child's `initState` or `didUpdateWidget` methods. How can I do it in the child itself without interfering with the state build ? Is there a better way to do this?
> 
> 1. How to combine state changes for multiple providers - In the example, you can see that `Service1` has a method called `incrementCounter` which also makes some updates in `Service2` as well. As I understand, each of these state changes in the services would notify a consumer in the widget tree. Is there a way to combine such changes together so that all the consumers are notified in one go ?
> 2. When to properly do Autodispose or dispose ? Is there a best practice for that ?

Can you please answer 1 ?

**Comment by rrousselGit** [comment]
In general, you shouldn't initialize providers inside "initState".
Inside initialize providers inside the providers themselves. 

If a provider initialization requires values from outside, store those values inside providers.
Then the initialized provider can read those values using `ref.watch`

**Comment by rrousselGit** [comment]
👋 
I'll close this since this was answered. Feel free to create a new issue if you still have the problem 😄
